test/integration/apiserver/admissionwebhook/match_conditions_test.go

```diff
@@ -676,6 +565,158 @@ func Test_MatchConditions(t *testing.T) {
+
+	testcases := []struct {
+		name            string
+		matchConditions []admissionregistrationv1.MatchCondition
+		expectError     bool
+	}{{
+		name: "valid match conditions",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       "exclude-leases",
+			Expression: "!(request.resource.group == 'coordination.k8s.io' && request.resource.resource == 'leases')",
+		}, {
+			Name:       "exclude-kubelet-requests",
+			Expression: "!('system:nodes' in request.userInfo.groups)",
+		}, {
+			Name:       "breakglass",
+			Expression: "!authorizer.group('admissionregistration.k8s.io').resource('validatingwebhookconfigurations').name('my-webhook.example.com').check('breakglass').allowed()",
+		}},
+		expectError: false,
+	}, {
+		name: "invalid field should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       "old-object-is-null.kubernetes.io",
+			Expression: "imnotafield == null",
+		}},
+		expectError: true,
+	}, {
+		name: "missing expression should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name: "old-object-is-null.kubernetes.io",
+		}},
+		expectError: true,
+	}, {
+		name: "missing name should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Expression: "oldObject == null",
+		}},
+		expectError: true,
+	}, {
+		name: "empty name should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       "",
+			Expression: "oldObject == null",
+		}},
+		expectError: true,
+	}, {
+		name: "empty expression should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       "test-empty-expression.kubernetes.io",
+			Expression: "",
+		}},
+		expectError: true,
+	}, {
+		name: "duplicate name should error",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       "test1",
+			Expression: "oldObject == null",
+		}, {
+			Name:       "test1",
+			Expression: "oldObject == null",
+		}},
+		expectError: true,
+	}, {
+		name: "name must be qualified name",
+		matchConditions: []admissionregistrationv1.MatchCondition{{
+			Name:       " test1",
+			Expression: "oldObject == null",
+		}},
+		expectError: true,
+	}}
+
+	endpoint := "https://localhost:1234/server"
+	for _, testcase := range testcases {
+		t.Run(testcase.name, func(t *testing.T) {
+			rules := []admissionregistrationv1.RuleWithOperations{{
+				Operations: []admissionregistrationv1.OperationType{admissionregistrationv1.Create},
+				Rule: admissionregistrationv1.Rule{
+					APIGroups:   []string{""},
+					APIVersions: []string{"v1"},
+					Resources:   []string{"pods"},
+				},
+			}}
+			clientConfig := admissionregistrationv1.WebhookClientConfig{
+				URL:      &endpoint,
+				CABundle: localhostCert,
+			}
+			versions := []string{"v1"}
+			validatingwebhook := &admissionregistrationv1.ValidatingWebhookConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "admission.integration.test",
```

Comment:
I'm surprised this test passes even though we're creating the same webhook object against the same instance of apiserver and not deleting them after each test case

Reply:
The mutating case doesn't use dry run though interestingly

Reply:
Sorry I got this test and `TestMatchConditions` mixed up. This test doesn't use dry run so I'm still curious why this passes with consecutive run.

Reply:
Ah shoot, it should use dry run. It passed because there was only a single valid case, and apparently validation runs before the check to see if an object already exists. I added a second valid test case, and also switched to dryrun.

Good catch.